common/network-common/src/main/java/org/apache/spark/network/util/PooledByteBufAllocatorWithMetrics.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.util;
+
+import java.util.Iterator;
+import java.util.List;
+
+import io.netty.buffer.PoolArenaMetric;
+import io.netty.buffer.PoolChunkListMetric;
+import io.netty.buffer.PoolChunkMetric;
+import io.netty.buffer.PooledByteBufAllocator;
+
+/**
+ * A {@link PooledByteBufAllocator} providing some metrics.
+ */
+public class PooledByteBufAllocatorWithMetrics extends PooledByteBufAllocator {
+
+  public PooledByteBufAllocatorWithMetrics(
+      boolean preferDirect,
+      int nHeapArena,
+      int nDirectArena,
+      int pageSize,
+      int maxOrder,
+      int tinyCacheSize,
+      int smallCacheSize,
+      int normalCacheSize) {
+    super(preferDirect, nHeapArena, nDirectArena, pageSize, maxOrder, tinyCacheSize,
+      smallCacheSize, normalCacheSize);
+  }
+
+  public long offHeapUsage() {
+    return sumOfMetrics(directArenas());
+  }
+
+  public long onHeapUsage() {
+    return sumOfMetrics(heapArenas());
+  }
+
+  private long sumOfMetrics(List<PoolArenaMetric> metrics) {
+    long sum = 0;
+    for (int i = 0; i < metrics.size(); i++) {
+      PoolArenaMetric metric = metrics.get(i);
```

Comment:
nit: it's better to use `Iterator` pattern here, as the input list may not be an indexed list and `list.get(i)` becomes `O(n)`.